src/settings/lib/RuleEditor/RulesEditor.js

```diff
@@ -292,130 +226,159 @@ class RulesEditor extends React.Component {
-          gutter: "rules-foldgutter",
+          gutter: 'rules-foldgutter',
           indicatorOpen: openTri,
           indicatorFolded: foldedTri,
         },
-        gutters: ["CodeMirror-linenumbers", "rules-foldgutter"],
+        gutters: ['CodeMirror-linenumbers', 'rules-foldgutter'],
       },
       code: this.props.code,
-    }
+    };
   }
 
   componentDidMount() {
-    this.cm = this.cmComponent.editor;
-    //set up hinting
+    this.cm = this.cmComponentRef.current.editor;
+
     rulesHint(Codemirror, this.props);
     initSubMenuDataFetching(Codemirror, this.props);
 
-    // prettymuch always show the auto-complete.
+    // pretty much always show the auto-complete
     this.cm.on('cursorActivity', this.showHelp);
     this.cm.on('endCompletion', this.showHelp);
-    // this.cm.on('focus', this.showHelp);
+  }
+
+  componentWillReceiveProps(nextProps) { // eslint-disable-line react/no-deprecated
+    const nextState = {};
+
+    if (nextProps.typeMapping && !isEqual(nextProps.typeMapping, this.props.typeMapping)) {
+      nextState.typeMapping = nextProps.typeMapping;
+    }
+
+    if (nextProps.policyMapping && !isEqual(nextProps.policyMapping, this.props.policyMapping)) {
+      nextState.policyMapping = nextProps.policyMapping;
+    }
+
+    if (nextProps.completionLists && !isEqual(nextProps.completionLists, this.props.completionLists)) {
+      nextState.completionLists = nextProps.completionLists;
+    }
+
+    if (!isEmpty(nextState)) {
+      this.setState(({ codeMirrorOptions }) => {
+        const newCodeMirrorOptions = cloneDeep(codeMirrorOptions);
+
+        return { codeMirrorOptions: Object.assign(newCodeMirrorOptions.mode, nextState) };
+      });
+    }
+
+    if (nextProps.code !== this.props.code) {
+      this.setState(() => ({ code: nextProps.code }));
+    }
+
+    if (nextProps.filter !== this.props.filter) {
+      this.filterRules(nextProps.filter);
+    }
   }
 
   componentDidUpdate() {
     this.clearErrors();
     this.cm.refresh();
-    if(this.props.errors && this.props.errors.length > 0){
-      this.props.errors.forEach((err) => {
-        this.renderError(err.line, err.message);
-      })
+
+    const { errors } = this.props;
+
+    if (!isEmpty(errors)) {
+      errors.forEach(error => this.renderError(error.line, error.message));
     }
   }
 
-  filterRules(filter) {
-    this.filteredSections.forEach((fs) => {
-      fs.clear();
-    });
+  filterRules = filter => {
+    this.filteredSections.forEach(filteredSection => filteredSection.clear());
 
-    if ( filter === '' ) {
+    if (filter === '') {
       return;
     }
 
     // scan rows with '#'
     const res = [];
-    const re = new RegExp(filter,'i');
+    const re = new RegExp(filter, 'i');
     let found = true;
     const rng = {};
-    this.cm.eachLine((lh) => {
-      // test rule line for filer string... if
-      if(/^\s*#/.test(lh.text)){
-        if(!re.test(lh.text)){
+
+    this.cm.eachLine(lh => {
```

Comment:
I would suggest this variable renaming because its difficult to understand what does "lh" mean.

Reply:
What do you propose?